File: backend/src/test/java/com/programmers/pcquotation/member/service/AuthServiceTest.java
package com.programmers.pcquotation.member.service;


import com.programmers.pcquotation.domain.customer.dto.CustomerSignupRequest;
import com.programmers.pcquotation.domain.customer.entity.Customer;
import com.programmers.pcquotation.domain.customer.exception.CustomerAlreadyExistException;
import com.programmers.pcquotation.domain.customer.exception.IncorrectLoginAttemptException;
import com.programmers.pcquotation.domain.customer.exception.PasswordMismatchException;
import com.programmers.pcquotation.domain.customer.repository.CustomerRepository;
import com.programmers.pcquotation.domain.member.dto.LoginRequest;
import com.programmers.pcquotation.domain.member.service.AuthService;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
@ActiveProfiles("test")
@SpringBootTest
public class AuthServiceTest {
    @Autowired
    private AuthService authService;

    @MockitoBean
    private CustomerRepository customerRepository;

    @MockitoBean
    private PasswordEncoder passwordEncoder;




    @Test
    public void signup_passwordMismatch() {
        CustomerSignupRequest customerSignupRequest = new CustomerSignupRequest(
                "user1",
                "1234",
                "1111",
                "홍길동",
                "test@test.com",
                "가장 좋아하는 음식은",
                "밥"
        );

        assertThrows(PasswordMismatchException.class, () -> authService.processSignup(customerSignupRequest));
    }

    @Test
    public void signup_customerUserNameAlreadyExist() {
        CustomerSignupRequest customerSignupRequest = new CustomerSignupRequest(
                "user1",
                "1234",
                "1234",
                "홍길동",
                "test@test.com",
                "가장 좋아하는 음식은",
                "밥"
        );

        Customer customer = customerSignupRequest.toCustomer();
        when(customerRepository.getCustomerByUsername(customerSignupRequest.getUsername())).thenReturn(Optional.of(customer));

        assertThrows(CustomerAlreadyExistException.class, () -> authService.processSignup(customerSignupRequest));
    }

    @Test
    public void signup_customerEmailAlreadyExist() {
        CustomerSignupRequest customerSignupRequest = new CustomerSignupRequest(
                "user1",
                "1234",
                "1234",
                "홍길동",
                "test@test.com",
                "가장 좋아하는 음식은",
                "밥"
        );

        Customer customer = customerSignupRequest.toCustomer();
        when(customerRepository.getCustomerByEmail(customerSignupRequest.getEmail())).thenReturn(Optional.of(customer));

        assertThrows(CustomerAlreadyExistException.class, () -> authService.processSignup(customerSignupRequest));
    }



    @Test
    public void login_usernameNotFound() {
        LoginRequest customerLoginRequest = new LoginRequest(
                "user1",
                "1234"
        );

        when(customerRepository.getCustomerByUsername("user1")).thenReturn(Optional.empty());

        assertThrows(IncorrectLoginAttemptException.class, () -> {
            authService.processLoginCustomer(customerLoginRequest);
        });

        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        assertNull(authentication);
    }

    @Test
    public void login_incorrectPassword() {
        LoginRequest customerLoginRequest = new LoginRequest(
                "user1",
                "1234"
        );

        Customer customer = Customer.builder()
                .username("user1")
                .password("1111")
                .build();

        when(customerRepository.getCustomerByUsername("user1")).thenReturn(Optional.of(customer));

        assertThrows(IncorrectLoginAttemptException.class, () -> {
            authService.processLoginCustomer(customerLoginRequest);
        });

        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        assertNull(authentication);
    }
}

File: backend/src/test/java/com/programmers/pcquotation/domain/member/service/AuthServiceTest.kt
package com.programmers.pcquotation.domain.member.service

import com.programmers.pcquotation.domain.customer.dto.CustomerSignupRequest
import com.programmers.pcquotation.domain.customer.entity.Customer
import com.programmers.pcquotation.domain.customer.exception.CustomerAlreadyExistException
import com.programmers.pcquotation.domain.customer.exception.IncorrectLoginAttemptException
import com.programmers.pcquotation.domain.customer.exception.PasswordMismatchException
import com.programmers.pcquotation.domain.customer.repository.CustomerRepository
import com.programmers.pcquotation.domain.member.dto.LoginRequest
import com.programmers.pcquotation.domain.member.service.AuthService
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.bean.override.mockito.MockitoBean
import java.util.*

@ActiveProfiles("test")
@SpringBootTest
class AuthServiceTest {
    @Autowired
    private lateinit var authService: AuthService

    @MockitoBean
    private lateinit var customerRepository: CustomerRepository

    @MockitoBean
    private lateinit var passwordEncoder: PasswordEncoder

    @Test
    fun signup_passwordMismatch() {
        val customerSignupRequest = CustomerSignupRequest(
            "user1",
            "1234",
            "1111",
            "홍길동",
            "test@test.com",
            "가장 좋아하는 음식은",
            "밥"
        )

        Assertions.assertThrows(PasswordMismatchException::class.java) {
            authService.processCustomerSignup(customerSignupRequest)
        }
    }

    @Test
    fun signup_customerUserNameAlreadyExist() {
        val customerSignupRequest = CustomerSignupRequest(
            "user1",
            "1234",
            "1234",
            "홍길동",
            "test@test.com",
            "가장 좋아하는 음식은",
            "밥"
        )

        val customer = customerSignupRequest.toCustomer()
        Mockito.`when`(customerRepository.getCustomerByUsername(customerSignupRequest.username))
            .thenReturn(Optional.of(customer))

        Assertions.assertThrows(CustomerAlreadyExistException::class.java) {
            authService.processCustomerSignup(customerSignupRequest)
        }
    }

    @Test
    fun signup_customerEmailAlreadyExist() {
        val customerSignupRequest = CustomerSignupRequest(
            "user1",
            "1234",
            "1234",
            "홍길동",
            "test@test.com",
            "가장 좋아하는 음식은",
            "밥"
        )

        val customer = customerSignupRequest.toCustomer()
        Mockito.`when`(customerRepository.getCustomerByEmail(customerSignupRequest.email))
            .thenReturn(Optional.of(customer))

        Assertions.assertThrows(CustomerAlreadyExistException::class.java) {
            authService.processCustomerSignup(customerSignupRequest)
        }
    }


    @Test
    fun login_usernameNotFound() {
        val customerLoginRequest = LoginRequest(
            "user1",
            "1234"
        )

        Mockito.`when`(customerRepository.getCustomerByUsername("user1")).thenReturn(Optional.empty())

        Assertions.assertThrows(IncorrectLoginAttemptException::class.java) {
            authService.processLoginCustomer(customerLoginRequest)
        }

        val authentication = SecurityContextHolder.getContext().authentication
        Assertions.assertNull(authentication)
    }

    @Test
    fun login_incorrectPassword() {
        val customerLoginRequest = LoginRequest(
            "user1",
            "1234"
        )

        val customer = Customer(
            "user1",
            "1111"
        )

        Mockito.`when`(customerRepository.getCustomerByUsername("user1")).thenReturn(Optional.of(customer))

        Assertions.assertThrows(IncorrectLoginAttemptException::class.java) {
            authService.processLoginCustomer(customerLoginRequest)
        }

        val authentication = SecurityContextHolder.getContext().authentication
        Assertions.assertNull(authentication)
    }
}